### PR TITLE
Add pickleshare and python-pathlib2.

### PR DESCRIPTION
**Pull request by pelson**
**PickleShare** - a small 'shelve' like datastore with concurrency support

Like shelve, a PickleShareDB object acts like a normal dictionary. Unlike shelve, many processes can access the database simultaneously. Changing a value in database is immediately visible to other processes accessing the same database.

-----

**python-pathlib2**

Fork of pathlib aiming to support the full stdlib Python API.

The old pathlib module on bitbucket is in bugfix-only mode. The goal of pathlib2 is to provide a backport of standard pathlib module which tracks the standard library module, so all the newest features of the standard pathlib can be used also on older Python versions.

### CODE REVIEW

**Comment by pelson** [comment]
Hi @mcmtroffaes and @vivainio, 

[conda-forge]( https://conda-forge.github.io/ ) is a community led packaging system for use with [`conda`]( http://conda.pydata.org/docs/ ), and I just wanted to let you know that we are adding `pathlib2` and `pickleshare` to the [list of packaged software on conda-forge](https://conda-forge.github.io/feedstocks.html).

The conda-forge infrastructure revolves around the idea of a *feedstock*, which is just a git repo containing a conda recipe (the build instructions) and automatically generated CI scripts necessary to build the recipe on Linux, OSX and Windows where appropriate. After merging this PR, `pathlib2` and `pickleshare` will each be given their own feedstock. The built conda distribution will be automatically uploaded to the to the [conda-forge channel](https://anaconda.org/conda-forge/) once the feedstock successfully runs its CI.

After an [auditing process (yet to be fully defined)]( https://github.com/conda-forge/conda-forge.github.io/issues/22#issuecomment-197117472 ), we hope packages submitted here will be included in the `defaults` channel distributed by Continuum, but in the meantime, everybody can benefit from these distributions by adding ``conda-forge`` as a conda channel.

We would love for you, and anybody else you can recommend, to be part of the [pathlib2-feedstock](http://github.com/conda-forge/python-pathlib2-feedstock/) and [pickleshare-feedstock](http://github.com/conda-forge/pickleshare-feedstock/) maintenance teams. Adding somebody to the team is simply a matter of adding your github handle to ```recipe-maintainers``` in the ```recipe/meta.yaml``` of the feedstock (i.e. it is part of the metadata itself) - conda-forge's automatic team management will take care of the rest and give you write access to the feedstock. Being part of the maintainer's team typically involves changing the version string when a new release of your software is made, responding to issues raised by users, as well as merging the occasional PR when automatic updates to the CI scripts are proposed.

If you'd like to be part of the feedstock maintainers list, but would rather not raise the PR yourself, just let me know and I'd be happy to update ```recipe-maintainers``` on your behalf.

File: recipes/python-pathlib2/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+about:
+  home: https://github.com/mcmtroffaes/pathlib2
+  license: MIT
+  summary: "Fork of pathlib aiming to support the full stdlib Python API."
+
+extra:
+  recipe-maintainers:
+    - pelson
+    - takluyver
```
Comment (pelson) [inline recipes/python-pathlib2/meta.yaml]: Meant to ping you on this @takluyver - I'm trying to increase the bus factor on the conda recipe so added you to the maintainers. Please do let me know if that isn't OK.

Comment (takluyver) [inline recipes/python-pathlib2/meta.yaml]: In the specific case, that's fine.

However, more generally, I'm concerned that we're replicating the Debian maintainership model, which I think is fundamentally broken for packaging a large diverse ecosystem of software components. Specifically, there are two related issues I have with it:

- If package maintainers are separate from upstream maintainers (as here - I'm not involved at all in pathlib2 development), I think things inevitably end up languishing because of a lack of interest. I'm unlikely to notice if a new version of pathlib2 is released, and even if I do, will I be motivated to update the conda recipe?
- If packages are maintained by specific named individuals rather than as a collective team pool, it means there's a low bus factor on each individual package, and it can lead to unwanted fragmentation, if different people pick different ways of packaging things.

I wonder if we could develop a more automated approach for straightforward packages, that wouldn't require human intervention on each new release. E.g. maybe convert wheels into conda packages? I have been working on a tool called [flonda](https://github.com/takluyver/flonda) that can build conda packages directly from the metadata my packaging tool [flit](http://flit.readthedocs.org/en/latest/) uses. That only works on packages that have chosen to use flit, but a similar technique could be used to make conda packages from wheels.

Comment (takluyver) [inline recipes/python-pathlib2/meta.yaml]: (Plus, even in the cases where I am the upstream maintainer, I'm not entirely keen on adding a step to my release process which has to happen in a separate repository.)

Comment (jakirkham) [inline recipes/python-pathlib2/meta.yaml]: > I wonder if we could develop a more automated approach for straightforward packages, that wouldn't require human intervention on each new release. E.g. maybe convert wheels into conda packages?

There has been discussed many times with Continuum over a couple of years on this point of simplifying the entry of PyPI packages into the conda ecosystem. Recently, it sounded like there was some interest on there end. Though I don't know what timeline is being considered at present. Really it seems like that is where this change will need to begin for anything to happen on this front.

Even if it does happen, there are some packages that are just problems for this model of packaging. At least to me, it is not entirely clear how this gets automated on their end. How do they recognize whether a package can be converted this way or not? For instance, pyzmq is on PyPI, but I think we would rather build its C dependencies and compile pyzmq to link against them. Similarly, numpy is up there with ATLAS, but many conda users would be extremely unhappy using ATLAS. Not to mention, this falls into the same category as pyzmq. Namely, we would want to build the BLAS dependency separately. They can try to read the PyPI package metadata for C/C++/Cython and skip it, but that only goes so far. For example, Python libraries using `ctypes` or `cffi` still have C library dependencies that we should try to package if we can. At present, the best we can do is what we are doing now. Wrap up the Python dependencies we need here. This is already a significant improvement on the previous situation given the level of automation present here.

Within our own system, we have considered at the bare minimum trying to open [issues]( https://github.com/conda-forge/conda-forge.github.io/issues/51 ) on PyPI releases or when git tags are pushed. Ideally, these would be PR's with version updates. However, there is still a bit of diversity in our system that makes this not totally easy to automate. While desirable, there are simply more pressing problems. For awhile that has been CI efficiency and automation. Now, it is starting to seem like documentation and examples. Both before and now, a pressing issue has been community engagement, review of additions, and migration of conda-recipes. While this is going well thus far, there is still lots of work to do on that front. I think it is still some time off before we really sit down and evaluate this problem. That being said, we remain aware that it exists.

TL;DR - Sure, but it is not that simple. While it is an issue, there are other problems that are more pressing at present.

**Comment by ocefpaf** [comment]
>  I'm concerned that we're replicating the Debian maintainership model, which I think is fundamentally broken for packaging a large diverse ecosystem of software components. 

I agree that we are replicating a Debian-like maintainership model, or any other Linux distro actually. However, I disagree that this model is fundamentally broken.

> If package maintainers are separate from upstream maintainers (as here - I'm not involved at all in pathlib2 development), I think things inevitably end up languishing because of a lack of interest. I'm unlikely to notice if a new version of pathlib2 is released, and even if I do, will I be motivated to update the conda recipe?

Sure, that can happen. But bare in mind that not all packages we have here can be found at PyPI. Even if they are not all PyPI packages maps directly to what a `conda` user might want. (See @jakirkham examples above plus any other package that we patch to fit a certain model and/or fix a certain problem, but it is not possible to send such a patch upstream: e.g.: how shapely uses `ctypes` to find the right `geos`).

> If packages are maintained by specific named individuals rather than as a collective team pool, it means there's a low bus factor on each individual package, and it can lead to unwanted fragmentation, if different people pick different ways of packaging things.

But we are a collective team pool, we just delegate individual tasks to named individuals :wink:
One of our main goals is to avoid the fragmentation that already exists in conda packaging (see the absurd amount of channels with redundant packages.).

In conda-forge we are tackling this problem with:
- community reviewed recipes;
- constantly improving/creating tools, like the linter, to check for recipe consistency;
- building the binaries using CI services to with standardized build environment.

We are happy to hear any suggestion on how to be more efficient.

PS: I understand that many Python developers choose PyPI as their main service to host their packages. I also understand that `conda` failed to find a good solution to install packages directly from PyPI. I would love to just `conda install <some-PyPI-package>`, but that can go awry if the package is not pure Python. Maybe `flint` and `flonta` can help us solve that problem.

**Comment by takluyver** [comment]
Sure, there are definitely some packages where extra work will be needed to make different kinds of packages, and this mechanism can handle more diverse software than PyPI packages. I'm not suggesting that this mechanism is unnecessary.

But I'm concerned that a substantial fraction of conda forge might end up being duplicated metadata from PyPI which needs to be updated when packages are released on PyPI. That's a waste of people's time, and practically, people will forget it or not bother with it, so we'll end up with many outdated packages - like the situation in Debian (though there are other reasons Debian packages get out of date as well).

I envisage something like this:

- We maintain a simple list of PyPI package names (in addition to the separate feedstock repos for more complex packages).
- Adding a package to that list implies that the project is pure Python, released as wheels, and we trust the upstream maintainers enough that we'd `pip install` a new version without much thought.
- A bot periodically (~daily) checks those packages on PyPI for new versions
- When a new version appears or a new package is added to the list, an automated job downloads the wheel, runs a tool which converts that to the set of conda packages for different platforms, and uploads them.
- There would be some kind of checks and monitoring, because a package that was previously pure Python might gain an extension module, or a new version might be released without a wheel, or something. "Errors should never pass silently"

**Comment by pelson** [comment]
I agree with pretty much everything you've said @takluyver. For better or worse, PyPY must currently be the canonical source of simple python packages if we are not to duplicate everything. Though [PEP516](https://www.python.org/dev/peps/pep-0516/) may very well help with reducing that need by defining the metadata agnostically.

Having said that, we are already in a position of having a huge amount of redundancy in terms of package metadata. Just looks at [shapely on anaconda.org](https://anaconda.org/search?q=shapely) - this is fuelled by at least 5 distinct recipes (https://github.com/conda/conda-recipes/tree/master/python/shapely, https://github.com/SciTools/conda-recipes-scitools/tree/master/shapely, https://github.com/ioos/conda-recipes/tree/57342ed77da3431949019e7e85d716a5927ec6ce/shapely, https://github.com/nbren12/conda-recipes/blob/master/shapely/, https://bitbucket.org/team_dtocean/shapely-win32-py27/src/c9f47cb892382d1a47e3fa45e30da021513d24fd/recipe/?at=default) and to my knowledge not a single one with a contribution from the primary author. The aim here is not to shame authors into maintenance of this information, it is to provide a canonical place for maintainers to come to curate their package if they so wish. If they don't, people will still want to be able to install their software through conda (let's not get onto the pip vs conda discussion 😉 ), and it is my belief that having a single yet forkable place for that information to live is a healthy thing.

It may be pipe dream, or just a matter of getting on and doing it, but I can envisage a situation where the metadata of a wheel/pypa.json is used to automatically propose a feedstock pull request which not only updates the version, but also other metadata such as the requirements and entry points. My argument is that as a package maintainer myself, I'd be quite content to see a pull request in my notifications which simply asked me to review some trivial changes and then hit merge. I'd even be prepared to have the occasional noise of pull requests which updated the versions that my package gets compiled against (i.e. a PR which updates the build to include python 3.6 or numpy 1.12).

I'm not sure we're disagreeing on an awful lot here, but I don't think the conversation belongs in a merged PR. @takluyver - are you interested in joining us for a hangout sometime to discuss further? We have one scheduled for Friday (https://github.com/conda-forge/conda-forge.github.io/issues/89) and I hope to have them on at least a fortnightly basis thereafter.

Cheers,

**Comment by takluyver** [comment]
> are you interested in joining us for a hangout sometime to discuss further? We have one scheduled for Friday

Sure! What time on Friday?

**Comment by pelson** [comment]
> Sure! What time on Friday?

I guess that link in conda-forge/conda-forge.github.io#89 didn't work 😢 ? I've updated it with the direct hangouts link, and the UTC time.

Would you mind adding it as an agenda item on that ticket? We are going to be discussing things which you guys at Jupyter have already covered, so if you are happy to lend your time I suspect you will be able to help with a number of items in the agenda. Otherwise, I'd be happy to schedule you in at half-past if you want to duck in specifically for this discussion.

Thanks,

**Comment by jakirkham** [comment]
On the PyPI point, my understand is Continuum had some script(s) that could kind of (and I'm using "kind of" very loosely here as I don't remember the details :smile:) scrape PyPI and build packages. Though IIRC that code is not quite in a usable state. I think @msarahan has access to it, but I don't know where this fits into his current todos or really any plans with it. Here is the [important part of the thread]( https://github.com/conda-forge/conda-forge.github.io/issues/28#issuecomment-202640484 ) that is generally discussing what to do with pure Python packages.

**Comment by takluyver** [comment]
Oh, I missed the link - I was scanning for something that looked like a time. I'm glad I asked, though, because the link says a different time to the issue (1430 UTC). I'll join at 1400 UTC (=3pm BST).

**Comment by jakirkham** [comment]
Yes, I believe 1400 UTC is what it is suppose to be.

**Comment by ChrisBarker-NOAA** [comment]
> Even if it does happen, there are some packages that are just problems for this model of packaging. At least to me, it is not entirely clear how this gets automated on their end. 

I propose a really simple (yet robust) approach:

1) Any package that mirrors a package on pypi (most python packages) have a recipe hand-built (maybe that's a simple run of conda-skeleton and test, but nevertheless, SOME hand work) for it, as we do now.

2) we maintain a list of packages that are straightforward builds of source from pypi

3) we have a utility that periodically scans this list of packages -- looks to see if there is an new version on PyPI, and if there is, updates the package. 

That's it, and I suspect virtually all pure python packages could be maintained this way, and a good fraction of packages with extensions, too. Ones with complex dependence on C libs, etc would not be amenable to this, but it would be nice to automate a good chunk.

NOTE: I was inspired to do this when Ned Batchelder  posted somewhere asking about whether he should be directly building conda packages for his coverage tool -- he's been publishing to PyPi but found users getting old versions when they used conda -- which he'd like to avoid, but he really didn't want to maintain conda packages himself, and, frankly, there was no easy way to do that! (until conda forge, anyway...)

**Comment by takluyver** [comment]
>  Any package that mirrors a package on pypi (most python packages) have a recipe hand-built (maybe that's a simple run of conda-skeleton and test, but nevertheless, SOME hand work) for it, as we do now.

This certainly would require the least change to what conda-forge does now. But it seems kind of ugly to have hundreds or thousands of individual repos duplicating PyPI metadata and only being updated by a bot.

**Comment by ChrisBarker-NOAA** [comment]
doesn't seem that *that* ugly to me :-)

What metadata are you concerned about? we don't duplicate all the READMEs and trove classifiers and all that anyway, do we?

**Comment by msarahan** [comment]
I think the metadata has to be duplicated at some level so that it populates appropriate fields in the metadata that gets baked into the package.  I don't think it's necessary to have it in the recipe meta.yaml, though.

Instead, for these recipes, I would propose that they have a different source URL, such as:

```
source:
    pypi: <pypi url>
```

This would go through a particular (not yet existing) logic section in conda build, much like conda-skeleton.  The actual build would be performed from the output of this section.

There is still the requirement to define the package name and version.  maybe some jinja2 magic can make the version come from pypi rather than being hardcoded - or maybe we could even feed in the complete list of versions on pypi, so that people have whatever strict versions they might need?

**Comment by jakirkham** [comment]
We could always take a look at [these templates]( https://github.com/conda/conda-build/blob/master/conda_build/templates/setuptools.yaml ) and adapt them for conda-forge.

I have to agree though. The metadata is quite minimal here: license, summary, home, and checksum. All of which can be extracted from PyPI quite easily.

I do like the concept of have PyPI packages as automatically maintained feedstocks, @ChrisBarker-NOAA. Though I might want to discuss some modifications in the details. The big picture seems about right. At some point, we have to be able to adapt for packages that start including code that requires compilation. Having them as feedstocks already is perfect for that.

**Comment by janschulz** [comment]
I think a bot which submits PRs for packages, which basically are a pure python package (and leaves in the dependency chain) directly from PyPI would be great. But I wouldn't automate this to the point that packages get autobuild as soon as someone uploads something to pypi.

IMO the benefit of using a linux distro is that there is another layer of expertise between the package author and the user and the whole system is integrated (=what is the debian packaging guide). In some cases something "negative" is added (debians openssl fiasco) and sometimes different goals lead to frustration (e.g. debians "it's outdated" reputation), but in most cases it takes away the frustration setting up a system by hand. If I want to have "newer" packages and know what I'm doing, I can still `conda install pip; pip install ...`.

**Comment by ChrisBarker-NOAA** [comment]
hmm, I'm not sure -- I guess a recipe for "the latest version on PyPi" would be good, but I kind of like having the version baked in -- otherwise kind of hard to build a not-the-latest package.

Also -- while you are at it, it would be nice if conda recipes could be more DRY -- you often have the version string repeated in three or more places, for instance.

-CHB

**Comment by takluyver** [comment]
> doesn't seem that that ugly to me :-)

It could be worse, but using an entire git repo per package when all that's really needed is a name and some way record of what releases have already been packaged.

> What metadata are you concerned about? we don't duplicate all the READMEs and trove classifiers and all that anyway, do we?

Looking at the recipes in this PR, beyond the name and version, we have:

- A download URL
- A filename to save it as (not sure why this even matters? it's just getting unpacked anyway)
- MD5 sum
- The build command (same in each recipe)
- Requirements and build requirements
- Import name (admittedly the provides field is rarely in PyPI metadata; packages built with flit do specify it)
- A home page URL
- License
- Summary

I'd like to be able to get rid of all of those.

> I think the metadata has to be duplicated at some level so that it populates appropriate fields in the metadata that gets baked into the package. I don't think it's necessary to have it in the recipe meta.yaml, though.

I think this is what I'm saying - you need that metadata in the packages that get built, and that's fine. It's maintaining copies of it all in these individual repos that strikes me as a bad idea.

**Comment by ChrisBarker-NOAA** [comment]
I like the template idea -- rather than my kludgy idea of a recipe that gets updated in place, you'd have a template for a recipe, and it would get re-processed each time.

**Comment by jakirkham** [comment]
Exactly, it would be like doing some sort of re-rendering each time, @ChrisBarker-NOAA.

@takluyver, take a look at this [jinja template for setuptools]( https://github.com/conda/conda-build/blob/master/conda_build/templates/setuptools.yaml ). It basically convinces the package to inject its metadata right into the template. Even though there would be a full repo. The underlying mechanism would be the working list of PyPI packages for this strategy.

Using some sort of webhook (like what I understand @janschulz to be suggesting), this process would be completely automated. If we find that something doesn't work for this, we can always remove it from the list and start maintaining it manually. This process would be extremely simple.

It would also allow us to easily add packages to automated maintenance. We could even have a special maintainer name for this exact process. Once this special named maintainer is added the recipe switches over to automated control. So, there wouldn't even be an explicit list per se. Though we could always generate one to see what is maintained.

**Comment by ChrisBarker-NOAA** [comment]
@takluyver:

when I started in all this (building up a channel of packages needed for my project), the first thing I did was  hack together a script that would essentially run conda skeleton, then conda build, then upload that to binstar.

so all I needed was a list of packages, and one command, and wham: the latest versions of everything. It sounds like you're proposing going tha way.

However, what I found was that:

- More packages than you'd think didn't simply build out of teh box.
- After the first push, I wasn't really adding many packages, so simply running conda skeleton, testing and then adding the recipe to my repo was fine, and pretty much the same amount of work I had to do anyway, since recipes seemed to always need tweeks.

However, it is tedious to update recipes for new versions, so hence my proposal.

On the other hand:

Conda forge is doing a LOT of packages

conda skeleton and conda build may have gotten better
we can influence the development of 

So maybe tweaking the tools, rather than tweaking the recipes is the better long term strategy.

semi-OT: conda skeleton should really use:

pip install ./

rather than:

setup.py install

That way you get the proper meta-data, it works with raw distutils packages, and you are less likely to trigger those damn easy-install dependency installations.

(maybe need a -no-deps flag, too..)

**Comment by ChrisBarker-NOAA** [comment]
On Wed, Apr 13, 2016 at 10:09 AM, Jan Schulz <notifications@github.com>
wrote:

> IMO the benefit of using a linux distro is that there is another layer of
> expertise between the package author and the user and the whole system is
> integrated
>
hmm -- that is what "Anaconda" provides - maybe conda-forge can be more raw?

-CHB

-- 

Christopher Barker, Ph.D.
Oceanographer

Emergency Response Division
NOAA/NOS/OR&R            (206) 526-6959   voice
7600 Sand Point Way NE   (206) 526-6329   fax
Seattle, WA  98115       (206) 526-6317   main reception

Chris.Barker@noaa.gov

**Comment by takluyver** [comment]
> take a look at this jinja template for setuptools.

Yikes. I'd like to make something simpler, not add another level of complexity on top of all the ones that are already there.

@ChrisBarker-NOAA I think there are a couple of big shifts underway in the Python ecosystem that are making that approach more and more appealing.

Firstly, as packaging gets more reliable, it's more appealing to split out smaller pieces into separate packages, and express dependencies between them. Projects are no longer proudly badged as having 'no dependencies except Python'. Jupyter/IPython is a prime example - it went from one big package with some bundled third-party libs to about a dozen packages and no (Python) bundling. This proliferation means there's more and more pieces to be repackaged, so more and more work if that involves manual maintenance.

However, because of this proliferation, and the shift to wheels, packages are becoming less quirky and more standardised. This makes automatic repackaging more practical. This is why I suggested limiting it to packages that publish wheels: wheels are standardised, and they don't get to run code at install time, so we know where the metadata is and where the files from the wheel need to go, and it should be relatively reliable to turn that into a conda package.

**Comment by ChrisBarker-NOAA** [comment]
@takluyver:

so how would this work? the build  script becomes:

pip install pacakge_name

??

and when you say wheel, do you mean binary wheel? or pure-python-only wheels?

-CHB

**Comment by takluyver** [comment]
> so how would this work? the build script becomes:
> pip install pacakge_name

You could do it like that. I'm thinking of a more drastic step, though: build a tool that can directly convert wheels to conda packages, without going through `conda-build`. Flonda already does something very similar, except the idea there is:

```
flit metadata -> wheel  (using flit)
flit metadata -> conda package  (using flonda)
```

Whereas I'm now proposing `wheel -> conda package`

Conda convert also does something like this, IIUC: 
```
conda package -> conda package for another platform
```

When you just need to rearrange files, conda-build is overkill. And this approach lets you make packages for all platforms easily, because we know how to arrange files for a Python package on Windows even if we're not running on Windows.

> and when you say wheel, do you mean binary wheel? or pure-python-only wheels?

Initially, I'd focus on pure-Python wheels, because that's what there's going to be most of, and we already have code to do similar things (flonda, conda-convert). But if the compilation options are compatible, it could be extended to binary wheels later.

**Comment by ChrisBarker-NOAA** [comment]
On Wed, Apr 13, 2016 at 11:12 AM, Thomas Kluyver <notifications@github.com>
wrote:

> When you just need to rearrange files, conda-build is overkill.
>
> fair enough.


> Initially, I'd focus on pure-Python wheels, because that's what there's
> going to be most of, and we already have code to do similar things (flonda,
> conda-convert). But if the compilation options are compatible, it could be
> extended to binary wheels later.
>
>
except binaries need to have linking paths re-written and such -- I"d let
conda-build take care of that for you.

-CHB


-- 

Christopher Barker, Ph.D.
Oceanographer

Emergency Response Division
NOAA/NOS/OR&R            (206) 526-6959   voice
7600 Sand Point Way NE   (206) 526-6329   fax
Seattle, WA  98115       (206) 526-6317   main reception

Chris.Barker@noaa.gov

**Comment by takluyver** [comment]
> except binaries need to have linking paths re-written and such -- I"d let conda-build take care of that for you.

Fair enough - I readily profess my ignorance of compilers and shared libraries. It's really the pure-Python packages I'm interested in.

**Comment by mwcraig** [comment]
@takluyver @ChrisBarker-NOAA  -- I'll admit I've only skimmed this but https://github.com/astropy/conda-build-tools might be of interest for part of it. Essentially it is a script for making recipes with conda skeleton then building when possible, a generating a recipe from a template.

It looks in a yaml file for a list of packages and versions to be built. Though I haven't used it this feature, it will, if the version number is left out, ask pypi for the most recent version and try to build it.

Example builder repos are at https://github.com/astropy/conda-channel-astropy and https://github.com/openastronomy/conda-channel

I don't let it run without version numbers because I've had a couple of pull requests that were intended to build one or two packages suddenly balloon.
